### PR TITLE
[SPARK-38335][SQL] Implement parser support for DEFAULT column values

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Implement parser changes needed to support for DEFAULT column values as tracked in https://issues.apache.org/jira/browse/SPARK-38334.

Note that these are the parser changes only. Analysis support will take place in a following PR.

Background: in the future, CREATE TABLE and ALTER TABLE invocations will support setting column default values for later operations. Following INSERT, UPDATE, MERGE statements may then reference the value using the DEFAULT keyword as needed.

Examples:
```sql
CREATE TABLE T(a INT, b INT NOT NULL);

-- The default default is NULL
INSERT INTO T VALUES (DEFAULT, 0);
INSERT INTO T(b)  VALUES (1);
SELECT * FROM T;
(NULL, 0)
(NULL, 1)

-- Adding a default to a table with rows, sets the values for the
-- existing rows (exist default) and new rows (current default).
ALTER TABLE T ADD COLUMN c INT DEFAULT 5;
INSERT INTO T VALUES (1, 2, DEFAULT);
SELECT * FROM T;
(NULL, 0, 5)
(NULL, 1, 5)
(1, 2, 5) 
```

### Why are the changes needed?

This new API helps users run DDL and DML statements to add new values to tables and scan existing values from tables more easily.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test coverage in DDLParserSuite.scala

### CODE REVIEW

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (gengliangwang) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:139]: I think we need to update line 149 `replaceTableHeader ('(' colTypeList ')')? tableProvider?` as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3494]: This method is override in the sq/core module https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala#L303
We need to update it as well

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3501]: Thanks for pointing this out! I had forgot to run the relevant subset of unit tests. Fixed this.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1427]: this method is not used

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1438]: this method is not used

**Review by gengliangwang — APPROVED** [review]
LGTM except minor comments

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1427]: Done. (I had considered refactoring some MERGE INTO unit testing to share the logical plans, but it turned out to be more confusing than expected, so I skipped that approach.)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1438]: Done.

**Comment by dtenedor** [comment]
jenkins merge

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2790]:
```suggestion
    if (defaultExpr.isDefined) {
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1451]:
```suggestion
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3874]:
```suggestion
        if (Option(colType.defaultExpression()).map(visitDefaultExpression).isDefined) {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3709]:
```suggestion
    if (defaultExpr.isDefined) {
```

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (HyukjinKwon) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1994]:
```suggestion
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2791]: Should we leverage our error framework e.g., in `QueryParsingErrors`? cc @MaxGekk FYI

**Comment by gengliangwang** [comment]
Before merging this PR, I would like to collect the opinions of more committers.
Supporting default column values is very common among DBMS. However, this will be a breaking change for Spark SQL
Currently Spark SQL
```
> create table t(i int, j int);
> insert into t values(1);
Error in query: `default`.`t` requires that the data to be inserted have the same number of columns as the target table: target table has 2 column(s) but the inserted data has 1 column(s), including 0 partition column(s) having constant value(s).
```

After supporting default column value:
```
> create table t(i int, j int);
> insert into t values(1);
> select * from t;
1	NULL
```

cc @cloud-fan  @dongjoon-hyun @viirya @dbtsai @huaxingao @maropu @zsxwing @wangyum @yaooqinn WDYT?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for pinging me, @gengliangwang .

**Comment by dongjoon-hyun** [comment]
cc @aokolnychyi , @RussellSpitzer , @rdblue , too.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (viirya) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1094]: Hmm, what is this for?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:2240]: nit: support -> parser support?

**Review by viirya — COMMENTED** [review]
Thanks for the work.

For the parser change itself, looks okay. As this is a breaking change, I'd like to see some clarification on why this is necessary to have. What issue we have if we don't have this (because we don't have the default value for long time) and do we have any workaround now?

**Review by dongjoon-hyun — COMMENTED** [review]
This PR doesn't seem to have the full body yet, what is your release target for this, @dtenedor and @gengliangwang ? I'm curious about the general error handling.
- Creating `NULL` default value for `NOT NULL` column
- Type mismatch between `default value literal` and column type.
- Upcasting or not in case of type mismatch

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (wangyum) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1093]: Do other databases support `SET DEFAULT expression`?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (gengliangwang) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1094]: For supporting the ANSI SQL syntax
```
ALTER TABLE table_name ALTER COLUMN column_name SET DEFAULT expr

ALTER TABLE table_name ALTER COLUMN column_name DROP DEFAULT
```

**Comment by gengliangwang** [comment]
> As this is a breaking change, I'd like to see some clarification on why this is necessary to have. 

@viirya Column default support is very helpful for table schema evolution and DML operations on wide tables with sparse data.

**Comment by HyukjinKwon** [comment]
Adding the default support itself should be fine - it's pretty common up to my best knowledge. My only concern is the breaking behavior (https://github.com/apache/spark/pull/35690#issuecomment-1057620996). I am fine w/ this as long as this matches w/ most of DBMSes. I suspect it's less critical since it failed before but now works.

**Comment by gengliangwang** [comment]
> Creating NULL default value for NOT NULL column
Type mismatch between default value literal and column type.
Upcasting or not in case of type mismatch

IMO:
- Not Null column can't have Null default
- Type mismatch between default value literal and column type:  we can simply forbid this. Note that we have many numeric types(Byte/Short/Int/Long/Decimal/Float/Double). If both default value literal type and column type are Numeric, it is not considered a mismatch.
- Upcasting or not in case of type mismatch: casting can happen if both of the literal type and column type are Numeric

@dtenedor WDYT?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
Comment (gengliangwang) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1093]: Yes, it is part of the ANSI SQL.
image

**Comment by gengliangwang** [comment]
> what is your release target for this

@dongjoon-hyun This is for Spark 3.4

**Comment by dongjoon-hyun** [comment]
Thank you for the confirmation.

**Comment by viirya** [comment]
> Adding the default support itself should be fine - it's pretty common up to my best knowledge. My only concern is the breaking behavior ([#35690 (comment)](https://github.com/apache/spark/pull/35690#issuecomment-1057620996)). I am fine w/ this as long as this matches w/ most of DBMSes. I suspect it's less critical since it failed before but now works.

I agreed with you. @HyukjinKwon 

And thanks for clarification, @gengliangwang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2790]: Done. Thanks for your reviews!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1451]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3874]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3709]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2791]: Sure, thanks for letting me know about this framework. Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:2240]: Done.

**Comment by dtenedor** [comment]
> Thanks for the work.
> 
> For the parser change itself, looks okay. As this is a breaking change, I'd like to see some clarification on why this is necessary to have. What issue we have if we don't have this (because we don't have the default value for long time) and do we have any workaround now?

This is for supporting the ANSI SQL syntax to specify default column values in DDL commands and then referencing them later in INSERT/UPDATE/MERGE commands. Examples:

CREATE TABLE T(a INT, b INT NOT NULL);

-- The default default is NULL
INSERT INTO T VALUES (DEFAULT, 0);
INSERT INTO T(b)  VALUES (1);
SELECT * FROM T;
(NULL, 0)
(NULL, 1)

-- Adding a default to a table with rows, sets the values for the
-- existing rows (exist default) and new rows (current default).
ALTER TABLE T ADD COLUMN c INT DEFAULT 5;
INSERT INTO T VALUES (1, 2, DEFAULT);
SELECT * FROM T;
(NULL, 0, 5)
(NULL, 1, 5)
(1, 2, 5)

**Comment by dtenedor** [comment]
> This PR doesn't seem to have the full body yet, what is your release target for this, @dtenedor and @gengliangwang ? I'm curious about the general error handling.
> 
> * Creating `NULL` default value for `NOT NULL` column
> * Type mismatch between `default value literal` and column type.
> * Upcasting or not in case of type mismatch

These are good questions.

* The future plan for `NOT NULL` columns is to require a `DEFAULT` value. If no `DEFAULT` value is provided, the engine will throw an error. We can also return an error in the case of a provided `DEFAULT NULL` value for a `NOT NULL` column.
* The engine can perform an explicit type coercion from the provided type to the required type, e.g. integer to floating-point. If these types are not coercible in this way, the engine can return an error, e.g. floating-point to boolean.

**Comment by dtenedor** [comment]
> > Creating NULL default value for NOT NULL column
> > Type mismatch between default value literal and column type.
> > Upcasting or not in case of type mismatch
> 
> IMO:
> 
> * Not Null column can't have Null default
> * Type mismatch between default value literal and column type:  we can simply forbid this. Note that we have many numeric types(Byte/Short/Int/Long/Decimal/Float/Double). If both default value literal type and column type are Numeric, it is not considered a mismatch.
> * Upcasting or not in case of type mismatch: casting can happen if both of the literal type and column type are Numeric
> 
> @dtenedor WDYT?

Good questions, I replied above earlier. We can perform a type coercion from the provided type to the required type, or return an error if the types are not coercible in this way. We can use existing type coercion rules in the analyzer for this part for consistency with the rest of Spark. For example, coercing an integer to floating-point should work, but coercing a floating-point to boolean should return an error to the user.

**Review by dongjoon-hyun — COMMENTED** [review]
cc @MaxGekk too since he is a release manager for Apache Spark 3.3 who need to cut `branch-3.3`.

**Comment by dtenedor** [comment]
This is ready for another review round @gengliangwang @viirya @wangyum @HyukjinKwon @dongjoon-hyun :)

**Comment by dtenedor** [comment]
(Note the prior merge conflict in the lexer has now been resolved.)

**Review by HyukjinKwon — APPROVED** [review]
I'm good w/ this change.

**Comment by cloud-fan** [comment]
This is a parser-only change and the feature is not implemented yet, so definitely not a breaking change. But I'd like to confirm that, is every new SQL feature a breaking change? e.g. adding a new SQL function means that a query failed with "function not found" before, now succeeds. This doesn't seem like a breaking change to me. The same applies to accepting more parameters in a SQL function, accepting more parameter types, etc.

The code change itself LGTM.

**Comment by gengliangwang** [comment]
@dtenedor Thanks for the first contribution!
@HyukjinKwon @dongjoon-hyun @viirya  @wangyum @cloud-fan thanks for the inputs! I am merging this parser-only PR to unblock @dtenedor's works in this feature.

**Comment by dongjoon-hyun** [comment]
@gengliangwang and @cloud-fan . Why do we have Spark 3.4 patch in `master` branch for Apache Spark 3.3?
> @dongjoon-hyun This is for Spark 3.4

Are we going to revert this from `branch-3.3`? cc @MaxGekk

**Comment by dongjoon-hyun** [comment]
As I mentioned https://github.com/apache/spark/pull/35690#pullrequestreview-899459801, I assumed we are going to wait until @MaxGekk cut a branch-3.3.

**Comment by gengliangwang** [comment]
@dongjoon-hyun I am merging this one to unblock @dtenedor's work on the actual changes in catalogs.
I will:
* revert this one on branch-3.3
* make sure no new related PRs merged to master until branch-3.3 is cut

**Comment by MaxGekk** [comment]
> Are we going to revert this from branch-3.3? cc @MaxGekk

If there is a risk that it can hurt stability. Let's revert it. I will open a blocker for 3.3 than to not forget this.

**Comment by dongjoon-hyun** [comment]
Thank you for your confirmations, @gengliangwang and @MaxGekk !

**Comment by dongjoon-hyun** [comment]
Hey, @MaxGekk . Did you make a block issue?
To @gengliangwang and @HyukjinKwon , I saw the JIRA was resolved as 3.3 because we want to avoid our merge script shows `3.4` as a new version.
Screen Shot 2022-03-15 at 11 43 10 AM
 
Since today is the feature freeze day, I recover it to 3.4.

**Comment by HyukjinKwon** [comment]
👍

**Comment by MaxGekk** [comment]
> Hey, @MaxGekk . Did you make a block issue?

Here is the blocker SPARK-38566

**Pull request closed without merging**
